### PR TITLE
Implement transact on the vanchor contract

### LINKED ISSUE #16: Implement `transact`

### PR DESCRIPTION
**Pull request by dharjeezy**
closes #16 and #9

### CODE REVIEW

File: contracts/vanchor/lib.rs [hunk 1]
```diff
@@ -58,10 +94,38 @@ mod vanchor {
         HashError,
         /// Verify error
         VerifyError,
-        /// Edge already exists
+        /// Edege already exists
```
Comment (drewstone) [inline contracts/vanchor/lib.rs:97]:
```suggestion
        /// Edge already exists
```

File: contracts/vanchor/lib.rs [hunk 2]
```diff
@@ -118,6 +212,259 @@ mod vanchor {
...
+                    return Err(Error::InvalidDepositAmount);
+                };
+                if abs_ext_amt != recv_token_amt {
+                    return Err(Error::InsufficientFunds);
+                };
+            }
+
+            let fee_exists = ext_data_fee != 0;
+
+            if fee_exists {
```
Comment (drewstone) [inline contracts/vanchor/lib.rs:279]: If this function is exclusively for depositing then I don't see how a relayer would ever be participating or why there would be a fee. Also separately, fees are only ever taken in the wrapped token being deposited.

File: contracts/vanchor/lib.rs [hunk 3]
```diff
@@ -118,6 +212,259 @@ mod vanchor {
...
+
+                // TODO: execute token wrapper unwrap function
+            }
+
+            let fee_exists = ext_data_fee != 0;
+
+            if fee_exists {
+                if self
+                    .env()
+                    .transfer(ext_data.relayer.clone(), ext_data_fee)
```
Comment (drewstone) [inline contracts/vanchor/lib.rs:321]: This should not be a native token transfer. Recall that the submitter of this transaction is a _relayer_. That relayer should be receiving the wrapped token that is being withdrawn and unshielded from the shielded pool.

**Review by drewstone — CHANGES_REQUESTED** [review]
Looks good so far, but I don't see any actual token transfers occurring. Or wrap_deposit/withdraw_unwrap separation and implementation, which is OK if it's meant to be handled separately. Can you provide an update on how you're planning to implement this?

File: contracts/vanchor/lib.rs (hunk 2 above)
Comment (dharjeezy) [reply contracts/vanchor/lib.rs:279]: For this, I actually implemented this based on the protocol-cosmwasm implementation here(https://github.com/webb-tools/protocol-cosmwasm/blob/main/contracts/vanchor/src/contract.rs#L355)
But, I know you told me earlier to look closely with the protocol-solidity implementation too. Which was what I did with the token-wrapper implementation. 
I will take a very good look at the protocol-solidity side too for this vanchor implementation and revert.

File: contracts/vanchor/lib.rs (hunk 3 above)
Comment (dharjeezy) [reply contracts/vanchor/lib.rs:321]: Alright. I will revert on this. This vanchor contract is not done yet until token-wrapper contract is ready.

**Comment by dharjeezy** [comment]
> Looks good so far, but I don't see any actual token transfers occurring. Or wrap_deposit/withdraw_unwrap separation and implementation, which is OK if it's meant to be handled separately. Can you provide an update on how you're planning to implement this?

Yes, no token wrap/unwrap is occurring yet, that's why I took my time to implement the token-wrapper contract.
I will sure plug the token-wrapper contract in with the vanchor contract.

**Pull request merged**